File: client/src/bridge/command-sender.js
/**
 * CommandSender — 封装发送指令的便捷 API
 */
import { gameBridge } from './game-bridge.js';

/**
 * 发送建造/升级指令
 * @param {string} nodeId
 * @param {string} buildingType - FARM|MINE|ARSENAL|WALL|ORACLE_BEACON
 */
export function sendBuild(nodeId, buildingType) {
  gameBridge.sendCommand('BUILD', { nodeId, buildingType });
}

/**
 * 发送科技研发指令
 * @param {string} techId
 */
export function sendResearch(techId) {
  gameBridge.sendCommand('RESEARCH', { techId });
}

/**
 * 发送进攻指令
 * @param {string} fromNodeId
 * @param {string} targetNodeId
 * @param {number} troopCount
 */
export function sendAttack(fromNodeId, targetNodeId, troopCount) {
  gameBridge.sendCommand('ATTACK', { fromNodeId, targetNodeId, troopCount });
}

export function sendRetreat(armyId) {
  gameBridge.sendCommand('RETREAT', { armyId, troopCount: armyId });
}

export function sendCreateCompany(nodeId, unitDefId = 'MILITIA') {
  gameBridge.sendCommand('CREATE_COMPANY', { nodeId, unitDefId });
}

export function sendMoveUnit(entityId, targetNodeId) {
  gameBridge.sendCommand('MOVE_UNIT', { entityId, targetNodeId });
}

export function sendAttackNode(entityId, targetNodeId) {
  gameBridge.sendCommand('ATTACK_NODE', { entityId, targetNodeId });
}

export function sendRetreatUnit(entityId) {
  gameBridge.sendCommand('RETREAT_UNIT', { entityId, armyId: entityId, troopCount: entityId });
}

export function sendCreateFormation(name, formationType, entityIds) {
  gameBridge.sendCommand('CREATE_FORMATION', { name, formationType, entityIds });
}

/**
 * 设置游戏速度
 * @param {number} speed - 0=暂停, 1=正常, 2=二倍速, 5=五倍速
 */
export function sendSetSpeed(speed) {
  gameBridge.sendCommand('SET_SPEED', { speed });
}

/**
 * 升级道路为铁轨
 * @param {string} edgeId
 */
export function sendUpgradeEdge(edgeId) {
  gameBridge.sendCommand('UPGRADE_EDGE', { edgeId });
}

/**
 * 创建运输路线
 * @param {string} fromNodeId
 * @param {string} targetNodeId
 * @param {string} cargoType - FOOD|IRON|AMMO
 */
export function sendCreateRoute(routeOrFromNodeId, targetNodeId, cargoType = 'FOOD') {
  if (typeof routeOrFromNodeId === 'object') {
    const {
      fromNodeId,
      targetNodeId: target,
      cargoType: cargo = 'FOOD',
      transportType = 'PORTER',
      transportCount = 1,
      priority = 50,
      routeMode = 'MANUAL',
    } = routeOrFromNodeId;
    gameBridge.sendCommand('CREATE_ROUTE', {
      fromNodeId,
      targetNodeId: target,
      cargoType: cargo,
      transportType,
      transportCount,
      priority,
      routeMode,
    });
    return;
  }
  gameBridge.sendCommand('CREATE_ROUTE', {
    fromNodeId: routeOrFromNodeId,
    targetNodeId,
    cargoType,
    transportType: 'PORTER',
    transportCount: 1,
    priority: 50,
    routeMode: 'MANUAL',
  });
}

/**
 * 取消运输路线
 * @param {number} entityId
 */
export function sendCancelRoute(entityId) {
  gameBridge.sendCommand('CANCEL_ROUTE', { routeId: entityId, troopCount: entityId });
}

export function sendUpdateRoute(routeId, patch = {}) {
  gameBridge.sendCommand('UPDATE_ROUTE', { routeId, troopCount: routeId, ...patch });
}

export function sendSetRallyPoint(nodeId, policies) {
  gameBridge.sendCommand('SET_RALLY_POINT', { nodeId, policies });
}

export function sendClearRallyPoint(nodeId) {
  gameBridge.sendCommand('CLEAR_RALLY_POINT', { nodeId });
}

export function sendProduceTransport(nodeId, transportType, quantity = 1) {
  gameBridge.sendCommand('PRODUCE_TRANSPORT', { nodeId, transportType, quantity });
}

// ─── 新增：3D 建筑放置 & 城墙绘制命令 ───

/**
 * 自由放置建筑
 * @param {string} nodeId
 * @param {string} buildingType - FARM|MINE|ARSENAL|ORACLE_BEACON
 * @param {number} localX - 相对城市的局部 X
 * @param {number} localZ - 相对城市的局部 Z
 * @param {number} rotation - Y轴旋转角度（度）
 */
export function sendPlaceBuilding(nodeId, buildingType, localX, localZ, rotation = 0) {
  gameBridge.sendCommand('PLACE_BUILDING', { nodeId, buildingType, localX, localZ, rotation });
}

/**
 * 建造城墙段
 * @param {string} nodeId
 * @param {number} fromX - 起点 X（局部坐标）
 * @param {number} fromZ - 起点 Z
 * @param {number} toX - 终点 X
 * @param {number} toZ - 终点 Z
 */
export function sendBuildWall(nodeId, fromX, fromZ, toX, toZ) {
  gameBridge.sendCommand('BUILD_WALL', { nodeId, fromX, fromZ, toX, toZ });
}

/**
 * 拆除建筑
 * @param {string} nodeId
 * @param {number} buildingIndex
 */
export function sendDemolishBuilding(nodeId, buildingIndex) {
  gameBridge.sendCommand('DEMOLISH_BUILDING', { nodeId, buildingIndex });
}

/**
 * 拆除城墙段
 * @param {string} nodeId
 * @param {number} wallIndex
 */
export function sendDemolishWall(nodeId, wallIndex) {
  gameBridge.sendCommand('DEMOLISH_WALL', { nodeId, wallIndex });
}

File: U3D/open-world-game-unity/Assets/Scripts/Networking/CommandSender.cs
using System.Collections.Generic;

/// <summary>
/// CommandSender — 封装发送指令的便捷 API
/// 移植自 client/src/bridge/command-sender.js
/// </summary>
namespace Networking
{
    public class CommandSender
    {
        private readonly WebSocketClient _ws;

        public CommandSender(WebSocketClient ws)
        {
            _ws = ws;
        }

        // ─── 建造/升级 ───
        public void SendBuild(string nodeId, string buildingType) =>
            _ws.SendCommand("BUILD", new { nodeId, buildingType });

        public void SendPlaceBuilding(string nodeId, string buildingType, float localX, float localZ, float rotation = 0) =>
            _ws.SendCommand("PLACE_BUILDING", new { nodeId, buildingType, localX, localZ, rotation });

        public void SendBuildWall(string nodeId, float fromX, float fromZ, float toX, float toZ) =>
            _ws.SendCommand("BUILD_WALL", new { nodeId, fromX, fromZ, toX, toZ });

        public void SendDemolishBuilding(string nodeId, int buildingIndex) =>
            _ws.SendCommand("DEMOLISH_BUILDING", new { nodeId, buildingIndex });

        public void SendDemolishWall(string nodeId, int wallIndex) =>
            _ws.SendCommand("DEMOLISH_WALL", new { nodeId, wallIndex });

        // ─── 科技研发 ───
        public void SendResearch(string techId) =>
            _ws.SendCommand("RESEARCH", new { techId });

        // ─── 军事 ───
        public void SendAttack(string fromNodeId, string targetNodeId, int troopCount) =>
            _ws.SendCommand("ATTACK", new { fromNodeId, targetNodeId, troopCount });

        public void SendRetreat(string armyId) =>
            _ws.SendCommand("RETREAT", new { armyId });

        public void SendCreateCompany(string nodeId, string unitDefId = "MILITIA") =>
            _ws.SendCommand("CREATE_COMPANY", new { nodeId, unitDefId });

        public void SendMoveUnit(int entityId, string targetNodeId) =>
            _ws.SendCommand("MOVE_UNIT", new { entityId, targetNodeId });

        public void SendAttackNode(int entityId, string targetNodeId) =>
            _ws.SendCommand("ATTACK_NODE", new { entityId, targetNodeId });

        public void SendRetreatUnit(int entityId) =>
            _ws.SendCommand("RETREAT_UNIT", new { entityId });

        public void SendCreateFormation(string name, string formationType, int[] entityIds) =>
            _ws.SendCommand("CREATE_FORMATION", new { name, formationType, entityIds });

        // ─── 游戏控制 ───
        public void SendSetSpeed(int speed) =>
            _ws.SendCommand("SET_SPEED", new { speed });

        // ─── 交通 ───
        public void SendUpgradeEdge(string edgeId) =>
            _ws.SendCommand("UPGRADE_EDGE", new { edgeId });

        // ─── 物流 ───
        public void SendCreateRoute(string fromNodeId, string targetNodeId, string cargoType = "FOOD",
            string transportType = "PORTER", int transportCount = 1, int priority = 50, string routeMode = "MANUAL") =>
            _ws.SendCommand("CREATE_ROUTE", new { fromNodeId, targetNodeId, cargoType, transportType, transportCount, priority, routeMode });

        public void SendCancelRoute(int routeId) =>
            _ws.SendCommand("CANCEL_ROUTE", new { routeId });

        public void SendUpdateRoute(int routeId, IDictionary<string, object>? patch = null)
        {
            var payload = new Dictionary<string, object>
            {
                ["routeId"] = routeId,
                ["troopCount"] = routeId
            };

            if (patch != null)
            {
                foreach (var (key, value) in patch)
                    payload[key] = value;
            }

            _ws.SendCommand("UPDATE_ROUTE", payload);
        }

        // ─── 集结点 ───
        public void SendSetRallyPoint(string nodeId, object policies) =>
            _ws.SendCommand("SET_RALLY_POINT", new { nodeId, policies });

        public void SendClearRallyPoint(string nodeId) =>
            _ws.SendCommand("CLEAR_RALLY_POINT", new { nodeId });

        // ─── 运输工具生产 ───
        public void SendProduceTransport(string nodeId, string transportType, int quantity = 1) =>
            _ws.SendCommand("PRODUCE_TRANSPORT", new { nodeId, transportType, quantity });
    }
}
